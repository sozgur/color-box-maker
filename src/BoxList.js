import { useState } from "react";
import { v4 as uuidv4 } from "uuid";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";

const BoxList = () => {
    const INITIAL_STATE = [];
    const [boxes, setBoxes] = useState(INITIAL_STATE);

    const addBox = (newBox) => {
        setBoxes([...boxes, { ...newBox, id: uuidv4() }]);
    };

    const removeBox = (id) => {
        setBoxes(boxes.filter((box) => id !== box.id));
    };

    return (
        <>
            <div>
                <NewBoxForm addBox={addBox} />
            </div>
            <div>
                {boxes.map(({ id, backgroundColor, width, height }) => (
                    <Box
                        key={id}
                        id={id}
                        backgroundColor={backgroundColor}
                        width={width}
                        height={height}
                        remove={removeBox}
                    />
                ))}
            </div>
        </>
    );
};

export default BoxList;
