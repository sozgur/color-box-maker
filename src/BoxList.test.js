import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", () => {
    render(<BoxList />);
});

it("matches snapshot", () => {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it("can add new box", () => {
    const { getByLabelText, getByText } = render(<BoxList />);
    const widthInput = getByLabelText("Width:");
    const heightInput = getByLabelText("Height:");
    const backgroundInput = getByLabelText("Background Color:");

    fireEvent.change(widthInput, { target: { value: "10" } });
    fireEvent.change(heightInput, { target: { value: "10" } });
    fireEvent.change(backgroundInput, { target: { value: "purple" } });

    const addBoxButton = getByText("Create Box");
    fireEvent.click(addBoxButton);

    const removeButton = getByText("X");
    expect(removeButton).toBeInTheDocument();

    expect(removeButton.parentElement).toHaveStyle(`
        width: 10em;
        height: 10em;
        background-color: purple;
        `);
});

it("can remove a box", () => {
    const { getByLabelText, getByText } = render(<BoxList />);
    const widthInput = getByLabelText("Width:");
    const heightInput = getByLabelText("Height:");
    const backgroundInput = getByLabelText("Background Color:");

    fireEvent.change(widthInput, { target: { value: "10" } });
    fireEvent.change(heightInput, { target: { value: "10" } });
    fireEvent.change(backgroundInput, { target: { value: "purple" } });

    const addBoxButton = getByText("Create Box");
    fireEvent.click(addBoxButton);

    const removeButton = getByText("X");
    expect(removeButton).toBeInTheDocument();

    fireEvent.click(removeButton);

    expect(removeButton).not.toBeInTheDocument();
});
