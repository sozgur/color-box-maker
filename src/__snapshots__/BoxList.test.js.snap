// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="backgroundColor"
      >
        Background Color: 
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        type="text"
        value=""
      />
      <label
        for="width"
      >
        Width: 
      </label>
      <input
        id="width"
        name="width"
        type="text"
        value=""
      />
      <label
        for="height"
      >
        Height: 
      </label>
      <input
        id="height"
        name="height"
        type="text"
        value=""
      />
      <button>
         Create Box
      </button>
    </form>
  </div>
  <div />
</DocumentFragment>
`;
